### ISSUE TITLE
Idea: a distributed push notifications service

### ISSUE THREAD

**Issue by KOLANICH** [UnifiedPush/specifications#36]
The idea description resides [in this GitHub issue](https://github.com/open-source-ideas/ideas/issues/57).

**Comment by MayeulC** [comment]
Hi @KOLANICH,

I am not sure what you are aiming at with this issue, but this is more or less exactly what UnifiedPush aims to solve.

It is not exactly distributed in the sense that UnifiedPush still relies on a client-server paradigm, but it is distributed in the sense that anybody can create a UnifiedPush implementation, server, and distributor.

> I think Tox can be used to transport these notifications to devices instead of centralised services

Are you suggesting to transport UnifiedPush notification over Tox? That would make Tox a UnifiedPush distributor, per [our definitions](https://unifiedpush.org/users/distributors/). The Conversations Android XMPP client does exactly that over XMPP. It can be done, but requires an HTTP endpoint that does HTTP -> Tox.

> [Tox] has to maintain a lot of connections anyway (and the battery is drained extremily quick, at least when using Antox)

It would seem the more appropriate course of action would be to use UnifiedPush to send notifications to Antox, to wake it up so that it can re-establish its energy-intensive connections. This is what push notifications are for: avoiding the need to maintain persistent connections that use a lot of battery.

The [TRIfA](https://github.com/zoff99/ToxAndroidRefImpl) Android client supports UnifiedPush this way, if you install the [companion app](https://github.com/zoff99/tox_push_msg_app). It is listed in the list of [compatible applications](https://unifiedpush.org/users/apps/) on the UnifiedPush website.

I also note that usually tokens are not required for UnifiedPush, one could imagine a client sending UnifiedPush notifications when it wants to wake up another client, almost completely p2p.

**Comment by KOLANICH** [comment]
>It would seem the more appropriate course of action would be to use UnifiedPush to send notifications to Antox, to wake it up so that it can re-establish its energy-intensive connections. This is what push notifications are for: avoiding the need to maintain persistent connections that use a lot of battery.

Yeah, I know. The problem is it'd require an external server. Maybe one can set up a server on a wireless home router for this purpose, but when doing so Tox on a phone no longer becomes "serverless". If one has to rely only on one's phone, then battery drain is inevitable.

>The TRIfA Android client supports UnifiedPush this way, if you install the companion app. It is listed in the list of compatible applications on the UnifiedPush website.

Thank you for the info.

**Comment by MayeulC** [comment]
> I guess it can cause additional power drain, if this server is hosted locally on phones.

Power drain is not the main issue, it should be negligible. The main issue is exposing a publicly accessible HTTP endpoint, despite (CG)NAT your phone is usually behind, plus your carrier's firewall. There are no open/forwarded ports, the only way to "open" ports is usually to rely on an external "STUN" server, and that's ephemeral.

One could imagine distributing push notifications through a DHT or an overlay network like yggdrasil, and that's doable, but the UnifiedPush specification assumes HTTP everywhere as a baseline. You can still implement your distributor and make it return, for instance, an `i2p:` address, but most servers will not be compatible with that.
This is something a future IPv7 could address, but not much can be done regarding this.

**Comment by KOLANICH** [comment]
>Power drain is not the main issue, it should be negligible. The main issue is exposing a publicly accessible HTTP endpoint, despite (CG)NAT your phone is usually behind, plus your carrier's firewall

Now I get what you meant. You meant the public gateway to which URI can be generated, that can be used by servers wanting to send Push Notifications.

I meant not that, having a public gateway and using it defeats the purpose of "serverlessness" - not to depend on the central servers and so operate even if central servers go down. The idea was not to optimize Tox itself in terms of power consumption, but to untie other apps on a phone from centralized services used to deliver pushes. It was meant that the only way to send a Tox-based push is to have 2 Tox clients, one is on phone and another one on the server sending a push (so

>but most servers will not be compatible with that.

applies and is by design, one cannot make an app decentralized without building such capabilities into an app), not via HTTP at all, but using a local Tox impl on server to send push through p2p directly without any external gateways. The client maintains the connection to p2p and receives the push from p2p. It drains battery, but it is inevitable if phone end user uses fully p2p apps.
